Add bugprone-loop-variable-copied-then-modified clang-tidy check.

flowerhack opened the pull request:
Adds a clang-tidy check that alerts when a loop variable is copied and subsequently modified.

This is a bugprone pattern because the programmer in this case often assumes they are modifying the original value instead of a copy.

This warning can be suppressed by either converting the loop variable to a const ref, or by performing the copy explicitly inside the body of the loop.

Fix GH-155922

flowerhack committed:
Add bugprone-loop-variable-copied-then-modified clang-tidy check.

Adds a clang-tidy check that alerts when a loop variable is copied and
subsequently modified.

This is a bugprone pattern because the programmer in this case often
assumes they are modifying the original value instead of a copy.

This warning can be suppressed by either converting the loop variable to
a const ref, or by performing the copy explicitly inside the body of the
loop.

Fix GH-155922

zwuis commented on clang-tools-extra/clang-tidy/bugprone/LoopVariableCopiedThenModifiedCheck.cpp:
```diff
@@ -0,0 +1,80 @@
+                  anyOf(materializeTemporaryExpr(), IteratorReturnsValueType,
+                        NotConstructedByCopy, ConstructedByConversion)))))));
+  Finder->addMatcher(
+      traverse(TK_AsIs,
+               cxxForRangeStmt(hasLoopVariable(LoopVar.bind("loopVar")))
+                   .bind("forRange")),
```
`TK_AsIs` is the default traverse mode. We can write `Matcher` instead of `traverse(TK_AsIs, Matcher)`.

zwuis commented on clang-tools-extra/clang-tidy/bugprone/LoopVariableCopiedThenModifiedCheck.cpp:
```diff
@@ -0,0 +1,80 @@
+  if (Var->getBeginLoc().isMacroID())
+    return;
+  const auto *ForRange = Result.Nodes.getNodeAs<CXXForRangeStmt>("forRange");
+  if (copiedLoopVarIsMutated(*Var, *ForRange, *Result.Context))
+    return;
```
Can we inline this function?

zwuis commented on clang-tools-extra/clang-tidy/bugprone/LoopVariableCopiedThenModifiedCheck.cpp:
```diff
@@ -0,0 +1,80 @@
+                                               const CXXForRangeStmt &ForRange,
+                                               ASTContext &Context) {
+
+  std::string hintstring = "";
```
Use camel case. See <https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly>.

zwuis commented on clang-tools-extra/clang-tidy/bugprone/LoopVariableCopiedThenModifiedCheck.cpp:
```diff
@@ -0,0 +1,100 @@
+    const MatchFinder::MatchResult &Result) {
+  const auto *Var = Result.Nodes.getNodeAs<VarDecl>("loopVar");
+  if (Var->getBeginLoc().isMacroID())
+    return;
```
Can we handle this case?

```cpp
template <typename> struct S {
  // ...
};
S<int> arr[1];
void f() {
  for (S s : arr) {
    s.modify();
  }
}
```

You can refer to the implementation of `clang::Type::isUndeducedType()`.

zwuis commented on clang-tools-extra/clang-tidy/bugprone/LoopVariableCopiedThenModifiedCheck.cpp:
```diff
@@ -0,0 +1,100 @@
+  const auto *ForRange = Result.Nodes.getNodeAs<CXXForRangeStmt>("forRange");
+  if (copiedLoopVarIsMutated(*Var, *ForRange, *Result.Context))
+    return;
+}
```
Variable is not needed. We can discard the result.

zwuis reviewed (commented):
> Adressing GH-155922.

You can add `Fix GH-155922`or something like this to PR description so that the issue will be closed when the PR is merged. See <https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue>.

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/LoopVariableCopiedThenModifiedCheck.cpp:
```diff
@@ -0,0 +1,100 @@
+    if (isa<AutoType>(LoopVar.getType())) {
+      hintstring = "const auto&";
+    } else {
+      std::string CanonicalTypeStr =
```
```suggestion
      const std::string CanonicalTypeStr =
```

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/LoopVariableCopiedThenModifiedCheck.h:
```diff
@@ -0,0 +1,39 @@
+
```
```suggestion
```

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/LoopVariableCopiedThenModifiedCheck.cpp:
```diff
@@ -0,0 +1,100 @@
+
```
```suggestion
```

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -139,6 +139,14 @@ New checks
+  <clang-tidy/checks/bugprone/loop-variable-copied-then-modified>` check.
+
+  Detects when a loop variable is copied and then subsequently modified.
+  Suggests replacing such instances with either a const ref (to prevent the
```
One sentence is enough.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/loop-variable-copied-then-modified.rst:
```diff
@@ -0,0 +1,47 @@
+bugprone-loop-variable-copied-then-modified
+===========================================
+
+Warns when a loop variable is copied and subsequently modified.
```
Please synchronize with statement in Release Notes.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/loop-variable-copied-then-modified.rst:
```diff
@@ -0,0 +1,47 @@
+
+This warning can be suppressed in one of two ways:
+  - In cases where the programmer did not intend to create a copy, they can
+    convert the loop variable to a const reference. A FixIt message will
```
```suggestion
    convert the loop variable to a ``const`` reference. A fix-it message will
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/loop-variable-copied-then-modified.rst:
```diff
@@ -0,0 +1,47 @@
+.. option:: IgnoreInexpensiveVariables
+
+   When `true`, this check will only alert on types that are expensive to copy.
+   This will lead to fewer "false" positives, but will also overlook some
```
```suggestion
   This will lead to fewer false positives, but will also overlook some
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/loop-variable-copied-then-modified.rst:
```diff
@@ -0,0 +1,47 @@
+
+   When `true`, this check will only alert on types that are expensive to copy.
+   This will lead to fewer "false" positives, but will also overlook some
+   instances where there may be an actual bug.
```
Please add default value.

flowerhack commented:
Thanks for the feedback; I believe I've addressed all comments, save one comment from @zwuis which I requested clarification on.  Let me know if there's anything else.

nicovank reviewed (commented):
Thanks!

I'm not a fan of the FixIt often leading to invalid code. Suggesting `const` when the variable is modified in the loop **always** produces invalid code (this is the case for the highlighted example in the documentation): suggesting `const` makes little sense when we know/assume the variable is modified.

Even if switching to reference only, the check will always transform the behavior of user code if applied directly. Most checks perform non-behavior-changing transformations. Maybe the FixIt can be a note instead or removed?

I ran this check on CMake sources (medium-sized codebase), there are a few true positives but also many false positives. Many being strings that are manipulated before being re-inserted into a different container for example. I couldn't think of a solution to reduce those yet.

One idea maybe is to check if the copy is then _used_ after modification. If it is not, that's a strong signal the copy was unintended. But this seems tricky.

flowerhack commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/loop-variable-copied-then-modified-ignore-inexpensive.cpp:
```diff
@@ -0,0 +1,55 @@
+#include "Inputs/system-header-simulator/sim_map"
+#include "Inputs/system-header-simulator/sim_unordered_map"
+#include "Inputs/system-header-simulator/sim_vector"
+#include "Inputs/system-header-simulator/sim_algorithm"
```
`sim_vector` depends on `sim_initializer_list` (which was indirectly included via `sim_set`), so I've added `sim_initializer_list` in and removed the rest.

flowerhack commented:
@nicovank : Thanks for the thoughtful and detailed comments & review.  Replying to your toplevel comment here:

> Suggesting const when the variable is modified in the loop always produces invalid code (this is the case for the highlighted example in the documentation): suggesting const makes little sense when we know/assume the variable is modified.

I performed an experiment where I ran this check over the Chromium codebase.  After filtering out uninteresting code (`*_test.cc` files, etc), it returned about 200 results.  Of those results, I picked about 50 to try handling by applying the naive fix—"if I switch this to a const ref, does it still compile?"—and in 70% of cases, the naive fix compiled without incident.  (I believe that the disparity between "things that the ExprMutationAnalyzer think represents a mutation" vs "things that the `const` designation thinks represents a mutation" is due to the ExprMutationAnalyzer's more conservative semantics.)

The remaining cases, to my eye, seemed like complicated enough situations that they would in fact benefit from more thought/care from the programmer, and it seems fine that the fix-it won't work out-of-the-box for those situations (generally require a judgment call from the programmer to either switch the variable in the loop to a non-const reference, or make the copy explicit in the loop).

So, my hope with the fix-it would be to offer a simple suggestion that would work "most" of the time to minimize disruption to developers.

That does mean some fixes suggested by this will not necessarily be bugfixes but rather a style enforcement—"prefer const references or references over copies in loop variables"— and that does make this something of an opinionated check, to be sure.  And the `IgnoreInexpensiveVariables` toggle does exist for those who'd like some benefit from this check with a higher signal-to-noise ratio.  But I do think it's valuable to have a version of the check that's concerned about *all* variables, since we've observed errors of this type involving inexpensive variables.

> One idea maybe is to check if the copy is then used after modification. If it is not, that's a strong signal the copy was unintended. But this seems tricky.

I think such a modification would fail to catch many of the cases we care about, unfortunately.  A pretty common pattern is to iterate over some set of objects (performing a copy on each iteration), modify a field in each object, and then continue.  In this case the field isn't "used" again inside the body of that function.
> One idea maybe is to check if the copy is then used after modification. If it is not, that's a strong signal the copy was unintended. But this seems tricky.

I think such a modification would fail to catch many of the cases we care about, unfortunately.  A pretty common pattern is to iterate over some set of objects (performing a copy on each iteration), modify a field in each object, and then continue.  In this case the field isn't "used" again inside the body of that function.

flowerhack commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/loop-variable-copied-then-modified.cpp:
```diff
@@ -0,0 +1,79 @@
+// RUN: %check_clang_tidy -std=c++17-or-later %s bugprone-loop-variable-copied-then-modified %t
```
C++17 has class template argument deduction, and I thought it'd be a useful test case to see if the FixIt does the correct thing in that case.  I can change that or put that case in a separate file, though, if you'd prefer?  Just let me know.

vbvictor replied:
You should make 2 RUN commands for different C++ standarts.
Since you target generic C++, then first one should be
`--std=c++98,c++03,c++11,c++14`, another one `--std=c++17-or-later`.
Look for other test files that have multiple RUN commands.

I _think_ for `--std=c++17-or-later` RUN command specifying `-check-suffix=,CPP17` would do the trick, like in `clang-tools-extra/test/clang-tidy/checkers/cppcoreguidelines/avoid-goto.cpp`

flowerhack commented:
(Stepping out for a couple hours, but I'll be adding the tests very shortly after that & will add another comment here when this is ready for another round of review.  Thanks!)

flowerhack commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/loop-variable-copied-then-modified.rst:
```diff
@@ -0,0 +1,48 @@
+    cases.
+  - In cases where the intent is in fact to modify a copy, they may perform the
+    copy inside the body of the loop, and perform whatever operations they like
+    on that copy.
```
Let me know if it'd be OK to resolve this conversation; thanks!

flowerhack commented:
This should be ready for another round of review!

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -167,6 +167,12 @@ New checks
+  Detects when a loop variable is copied and then subsequently (possibly) modified
+  and suggests replacing with a reference or an explicit copy.
+
 - New :doc:`bugprone-derived-method-shadowing-base-method
```
Please keep alphabetical order (by check name) in this section.

flowerhack replied:
alphabetized the checks~

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -161,16 +161,27 @@ Improvements to clang-tidy
+
+  Finds derived class methods that shadow a (non-virtual) base class method.
+ 
+ - New :doc:`bugprone-derived-method-shadowing-base-method
```
Duplicated entry. Incorrect merge/rebase?

flowerhack replied:
Augh, yeah, bad rebase.  I've re-pushed and I believe it's deduplicated now.

flowerhack committed:
Merge branch 'main' into make-as-new-checkout

flowerhack commented on clang-tools-extra/clang-tidy/bugprone/LoopVariableCopiedThenModifiedCheck.cpp:
```diff
@@ -0,0 +1,95 @@
+      << FixItHint::CreateInsertion(EndLoc, "&");
+}
+
+} // namespace clang::tidy::bugprone
```
done!

flowerhack committed:
Merge branch 'main' into make-as-new-checkout

flowerhack committed:
Merge branch 'main' into make-as-new-checkout

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/LoopVariableCopiedThenModifiedCheck.cpp:
```diff
@@ -0,0 +1,95 @@
+              unless(hasInitializer(expr(hasDescendant(expr(
+                  anyOf(materializeTemporaryExpr(), IteratorReturnsValueType,
+                        NotConstructedByCopy, ConstructedByConversion)))))));
+  Finder->addMatcher(cxxForRangeStmt(hasLoopVariable(LoopVar.bind("loopVar")))
```
consider excluding system headers or implicit code like template instances, otherwise there could be conflicts on fixes

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/LoopVariableCopiedThenModifiedCheck.cpp:
```diff
@@ -0,0 +1,95 @@
+    return;
+  }
+
+  clang::SourceRange LoopVarSourceRange =
```
LoopVar**Type**SourceRange

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/LoopVariableCopiedThenModifiedCheck.cpp:
```diff
@@ -0,0 +1,95 @@
+      << LoopVar->getName();
+  diag(LoopVar->getLocation(), "consider making '%0' a reference",
+       DiagnosticIDs::Note)
+      << LoopVar->getName()
```
no need for getName, it's safer without it

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/loop-variable-copied-then-modified.rst:
```diff
@@ -0,0 +1,55 @@
+This code can be fixed in one of two ways:
+  - In cases where the programmer did not intend to create a copy, they can
+    convert the loop variable to a reference or a ``const`` reference. A
+    fix-note message will provide a naive suggestion of how to achieve this,
```
do not refer to those fixes as "fix-note", mention somewhere that check provide fixes that can be enabled with --fix-notes. Also mention that those fixes can break compilation.

PiotrZSL reviewed (commented):
Few comments, overall not bad.
